File: lib/src/test/java/net/mm2d/upnp/SsdpNotifyReceiverTest.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import net.mm2d.upnp.SsdpNotifyReceiver.NotifyListener;
import net.mm2d.upnp.SsdpServer.Address;
import net.mm2d.upnp.SsdpServerDelegate.Receiver;
import net.mm2d.util.NetworkUtils;
import net.mm2d.util.TestUtils;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatchers;

import java.io.IOException;
import java.net.InetAddress;
import java.net.InterfaceAddress;
import java.net.NetworkInterface;

import static org.hamcrest.Matchers.*;
import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

@RunWith(JUnit4.class)
public class SsdpNotifyReceiverTest {
    @Test
    public void setNotifyListener_受信メッセージが通知されること() throws Exception {
        final NetworkInterface networkInterface = NetworkUtils.getAvailableInet4Interfaces().get(0);
        final SsdpServerDelegate delegate = spy(new SsdpServerDelegate(mock(Receiver.class), Address.IP_V4, networkInterface));
        final InterfaceAddress interfaceAddress = TestUtils.createInterfaceAddress("192.0.2.2", "255.255.255.0", 16);
        doReturn(interfaceAddress).when(delegate).getInterfaceAddress();
        final SsdpNotifyReceiver receiver = spy(new SsdpNotifyReceiver(delegate));

        final ArgumentCaptor<SsdpRequest> captor = ArgumentCaptor.forClass(SsdpRequest.class);
        final NotifyListener listener = mock(NotifyListener.class);
        doNothing().when(listener).onReceiveNotify(captor.capture());
        receiver.setNotifyListener(listener);

        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin");
        final InetAddress address = InetAddress.getByName("192.0.2.2");
        receiver.onReceive(address, data, data.length);

        assertThat(captor.getValue().getUuid(), is("uuid:01234567-89ab-cdef-0123-456789abcdef"));
    }

    @Test
    public void onReceive_同一セグメントからのメッセージは通知する() throws Exception {
        final SsdpNotifyReceiver receiver = spy(new SsdpNotifyReceiver(Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces().get(0)));
        final InterfaceAddress address = TestUtils.createInterfaceAddress("192.0.2.1", "255.255.0.0", 24);
        doReturn(address).when(receiver).getInterfaceAddress();
        final NotifyListener listener = mock(NotifyListener.class);
        receiver.setNotifyListener(listener);
        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin");

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), data, data.length);

        verify(listener).onReceiveNotify(ArgumentMatchers.any(SsdpRequest.class));
    }

    @Test
    public void onReceive_Listenerがnullでもクラッシュしない() throws Exception {
        final SsdpNotifyReceiver receiver = spy(new SsdpNotifyReceiver(Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces().get(0)));
        final InterfaceAddress address = TestUtils.createInterfaceAddress("192.0.2.1", "255.255.0.0", 24);
        doReturn(address).when(receiver).getInterfaceAddress();
        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin");

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), data, data.length);
    }

    @Test
    public void onReceive_異なるセグメントからのメッセージは無視する() throws Exception {
        final SsdpNotifyReceiver receiver = spy(new SsdpNotifyReceiver(Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces().get(0)));
        final InterfaceAddress address = TestUtils.createInterfaceAddress("192.0.2.1", "255.255.0.0", 24);
        doReturn(address).when(receiver).getInterfaceAddress();
        final NotifyListener listener = mock(NotifyListener.class);
        receiver.setNotifyListener(listener);
        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin");

        receiver.onReceive(InetAddress.getByName("192.1.2.2"), data, data.length);

        verify(listener, never()).onReceiveNotify(ArgumentMatchers.any(SsdpRequest.class));
    }

    @Test
    public void onReceive_M_SEARCHパケットは無視する() throws Exception {
        final SsdpNotifyReceiver receiver = spy(new SsdpNotifyReceiver(Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces().get(0)));
        final InterfaceAddress address = TestUtils.createInterfaceAddress("192.0.2.1", "255.255.0.0", 24);
        doReturn(address).when(receiver).getInterfaceAddress();
        final NotifyListener listener = mock(NotifyListener.class);
        receiver.setNotifyListener(listener);

        final SsdpRequest message = new SsdpRequest();
        message.setMethod(SsdpMessage.M_SEARCH);
        message.setUri("*");
        message.setHeader(Http.HOST, Address.IP_V4.getAddressString());
        message.setHeader(Http.MAN, SsdpMessage.SSDP_DISCOVER);
        message.setHeader(Http.MX, "1");
        message.setHeader(Http.ST, SsdpSearchServer.ST_ALL);
        final byte[] data = message.getMessage().getMessageString().getBytes();

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), data, data.length);

        verify(listener, never()).onReceiveNotify(ArgumentMatchers.any(SsdpRequest.class));
    }

    @Test
    public void onReceive_ByeByeパケットは通知する() throws Exception {
        final SsdpNotifyReceiver receiver = spy(new SsdpNotifyReceiver(Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces().get(0)));
        final InterfaceAddress address = TestUtils.createInterfaceAddress("192.0.2.1", "255.255.0.0", 24);
        doReturn(address).when(receiver).getInterfaceAddress();
        final NotifyListener listener = mock(NotifyListener.class);
        receiver.setNotifyListener(listener);

        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-notify-byebye0.bin");

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), data, data.length);

        verify(listener).onReceiveNotify(ArgumentMatchers.any(SsdpRequest.class));
    }

    @Test
    public void onReceive_LocationとSourceが不一致のメッセージは無視する() throws Exception {
        final SsdpNotifyReceiver receiver = spy(new SsdpNotifyReceiver(Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces().get(0)));
        final InterfaceAddress address = TestUtils.createInterfaceAddress("192.0.2.1", "255.255.0.0", 24);
        doReturn(address).when(receiver).getInterfaceAddress();
        final NotifyListener listener = mock(NotifyListener.class);
        receiver.setNotifyListener(listener);
        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin");

        receiver.onReceive(InetAddress.getByName("192.0.2.3"), data, data.length);

        verify(listener, never()).onReceiveNotify(ArgumentMatchers.any(SsdpRequest.class));
    }

    @Test
    public void onReceive_IOExceptionが発生してもクラッシュしない() throws Exception {
        final SsdpNotifyReceiver receiver = spy(new SsdpNotifyReceiver(Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces().get(0)));
        final InterfaceAddress address = TestUtils.createInterfaceAddress("192.0.2.1", "255.255.0.0", 24);
        doReturn(address).when(receiver).getInterfaceAddress();
        doThrow(new IOException()).when(receiver).createSsdpRequestMessage(ArgumentMatchers.any(byte[].class), anyInt());
        final NotifyListener listener = mock(NotifyListener.class);
        receiver.setNotifyListener(listener);
        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin");

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), data, data.length);

        verify(listener, never()).onReceiveNotify(ArgumentMatchers.any(SsdpRequest.class));
    }

    @Test
    public void send_delegateがコールされる() {
        final SsdpServerDelegate delegate = mock(SsdpServerDelegate.class);
        final SsdpNotifyReceiver receiver = new SsdpNotifyReceiver(delegate);
        final SsdpMessage message = mock(SsdpMessage.class);

        receiver.send(message);

        verify(delegate, times(1)).send(message);
    }

    @Test
    public void isSameSegment() throws Exception {
        assertThat(
                SsdpNotifyReceiver.isSameSegment(
                        TestUtils.createInterfaceAddress("192.168.0.1", "255.255.255.0", 24),
                        InetAddress.getByName("192.168.0.255")),
                is(true));
        assertThat(
                SsdpNotifyReceiver.isSameSegment(
                        TestUtils.createInterfaceAddress("192.168.0.1", "255.255.255.128", 25),
                        InetAddress.getByName("192.168.0.255")),
                is(false));
        assertThat(
                SsdpNotifyReceiver.isSameSegment(
                        TestUtils.createInterfaceAddress("192.168.0.1", "255.255.255.0", 24),
                        InetAddress.getByName("192.168.1.255")),
                is(false));
        assertThat(
                SsdpNotifyReceiver.isSameSegment(
                        TestUtils.createInterfaceAddress("192.168.0.1", "255.255.254.0", 23),
                        InetAddress.getByName("192.168.1.255")),
                is(true));
    }
}

File: mmupnp/src/test/java/net/mm2d/upnp/cp/internal/server/SsdpNotifyReceiverTest.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.cp.internal.server

import com.google.common.truth.Truth.assertThat
import io.mockk.*
import net.mm2d.upnp.common.Http
import net.mm2d.upnp.common.SsdpMessage
import net.mm2d.upnp.common.internal.message.SsdpRequest
import net.mm2d.upnp.common.internal.server.Address
import net.mm2d.upnp.common.internal.server.Address.IP_V4
import net.mm2d.upnp.common.internal.server.SsdpServerDelegate
import net.mm2d.upnp.common.internal.thread.TaskExecutors
import net.mm2d.upnp.common.util.NetworkUtils
import net.mm2d.upnp.util.TestUtils
import net.mm2d.upnp.util.createInterfaceAddress
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import java.io.IOException
import java.net.InetAddress

@Suppress("TestFunctionName", "NonAsciiCharacters")
@RunWith(JUnit4::class)
class SsdpNotifyReceiverTest {
    private lateinit var taskExecutors: TaskExecutors

    @Before
    fun setUp() {
        taskExecutors = TaskExecutors()
    }

    @After
    fun tearDown() {
        taskExecutors.terminate()
    }

    @Test
    fun setNotifyListener_受信メッセージが通知されること() {
        val delegate =
            spyk(SsdpServerDelegate(taskExecutors, IP_V4, NetworkUtils.getAvailableInet4Interfaces()[0]))
        delegate.setReceiver(mockk(relaxed = true))
        val interfaceAddress = createInterfaceAddress("192.0.2.2", "255.255.255.0", 16)
        every { delegate.interfaceAddress } returns interfaceAddress
        val receiver = spyk(SsdpNotifyReceiver(delegate))

        val slot = slot<SsdpRequest>()
        val listener: (SsdpMessage) -> Unit = mockk()
        every { listener.invoke(capture(slot)) } answers { nothing }
        receiver.setNotifyListener(listener)

        val data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin")
        val address = InetAddress.getByName("192.0.2.2")
        receiver.onReceive(address, 0, data, data.size)

        assertThat(slot.captured.uuid).isEqualTo("uuid:01234567-89ab-cdef-0123-456789abcdef")
    }

    @Test
    fun onReceive_同一セグメントからのメッセージは通知する() {
        val receiver =
            spyk(SsdpNotifyReceiver(taskExecutors, Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces()[0]))
        val address = createInterfaceAddress("192.0.2.1", "255.255.0.0", 24)
        every { receiver.interfaceAddress } returns address
        val listener: (SsdpMessage) -> Unit = spyk { }
        receiver.setNotifyListener(listener)
        val data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin")

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), 0, data, data.size)

        verify(exactly = 1) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_filterでfalseになるとlistenerコールされない() {
        val receiver =
            spyk(SsdpNotifyReceiver(taskExecutors, Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces()[0]))
        val address = createInterfaceAddress("192.0.2.1", "255.255.0.0", 24)
        every { receiver.interfaceAddress } returns address
        val listener: (SsdpMessage) -> Unit = spyk { }
        receiver.setNotifyListener(listener)
        val data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin")
        receiver.setFilter { false }

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), 0, data, data.size)

        verify(inverse = true) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_Listenerがnullでもクラッシュしない() {
        val receiver =
            spyk(SsdpNotifyReceiver(taskExecutors, Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces()[0]))
        val address = createInterfaceAddress("192.0.2.1", "255.255.0.0", 24)
        every { receiver.interfaceAddress } returns address
        val data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin")

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), 0, data, data.size)
    }

    @Test
    fun onReceive_異なるセグメントからのメッセージは無視する() {
        val receiver =
            spyk(SsdpNotifyReceiver(taskExecutors, Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces()[0]))
        val address = createInterfaceAddress("192.0.2.1", "255.255.0.0", 24)
        every { receiver.interfaceAddress } returns address
        val listener: (SsdpMessage) -> Unit = spyk { }
        receiver.setNotifyListener(listener)
        receiver.setSegmentCheckEnabled(true)
        val data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin")

        receiver.onReceive(InetAddress.getByName("192.1.2.2"), 0, data, data.size)

        verify(inverse = true) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_M_SEARCHパケットは無視する() {
        val receiver =
            spyk(SsdpNotifyReceiver(taskExecutors, Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces()[0]))
        val address = createInterfaceAddress("192.0.2.1", "255.255.0.0", 24)
        every { receiver.interfaceAddress } returns address
        val listener: (SsdpMessage) -> Unit = spyk { }
        receiver.setNotifyListener(listener)

        val message = SsdpRequest.create()
        message.setMethod(Http.M_SEARCH)
        message.setUri("*")
        message.setHeader(Http.HOST, Address.IP_V4.ssdpAddressString)
        message.setHeader(Http.MAN, SsdpMessage.SSDP_DISCOVER)
        message.setHeader(Http.MX, "1")
        message.setHeader(Http.ST, SsdpSearchServer.ST_ALL)
        val data = message.message.getMessageString().toByteArray()

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), 0, data, data.size)

        verify(inverse = true) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_ByeByeパケットは通知する() {
        val receiver =
            spyk(SsdpNotifyReceiver(taskExecutors, Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces()[0]))
        val address = createInterfaceAddress("192.0.2.1", "255.255.0.0", 24)
        every { receiver.interfaceAddress } returns address
        val listener: (SsdpMessage) -> Unit = spyk { }
        receiver.setNotifyListener(listener)

        val data = TestUtils.getResourceAsByteArray("ssdp-notify-byebye0.bin")

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), 0, data, data.size)

        verify(exactly = 1) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_LocationとSourceが不一致のメッセージは無視する() {
        val receiver =
            spyk(SsdpNotifyReceiver(taskExecutors, Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces()[0]))
        val address = createInterfaceAddress("192.0.2.1", "255.255.0.0", 24)
        every { receiver.interfaceAddress } returns address
        val listener: (SsdpMessage) -> Unit = spyk { }
        receiver.setNotifyListener(listener)
        val data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin")

        receiver.onReceive(InetAddress.getByName("192.0.2.3"), 0, data, data.size)

        verify(inverse = true) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_sony_telepathy_serviceは無視する() {
        val receiver =
            spyk(SsdpNotifyReceiver(taskExecutors, Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces()[0]))
        val address = createInterfaceAddress("192.0.2.1", "255.255.0.0", 24)
        every { receiver.interfaceAddress } returns address
        val listener: (SsdpMessage) -> Unit = spyk { }
        receiver.setNotifyListener(listener)
        val data = TestUtils.getResourceAsByteArray("ssdp-notify-alive-telepathy.bin")

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), 0, data, data.size)

        verify(inverse = true) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_IOExceptionが発生してもクラッシュしない() {
        val receiver =
            spyk(SsdpNotifyReceiver(taskExecutors, Address.IP_V4, NetworkUtils.getAvailableInet4Interfaces()[0]))
        val address = createInterfaceAddress("192.0.2.1", "255.255.0.0", 24)
        every { receiver.interfaceAddress } returns address
        every { receiver.createSsdpRequestMessage(any(), any()) } throws IOException()
        val listener: (SsdpMessage) -> Unit = spyk { }
        receiver.setNotifyListener(listener)
        val data = TestUtils.getResourceAsByteArray("ssdp-notify-alive0.bin")

        receiver.onReceive(InetAddress.getByName("192.0.2.2"), 0, data, data.size)

        verify(inverse = true) { listener.invoke(any()) }
    }

    @Test
    fun send_delegateがコールされる() {
        val delegate: SsdpServerDelegate = mockk(relaxed = true)
        val receiver = spyk(SsdpNotifyReceiver(delegate))
        val message: SsdpMessage = mockk(relaxed = true)

        receiver.send { message }

        verify(exactly = 1) { delegate.send(any()) }
    }
}
